src/coord/src/persistence.rs

```diff
@@ -0,0 +1,337 @@
+            let mut prefix_start_offset = None;
+
+            for p in partition.pending.iter() {
+                if prefix_start_offset.is_none() {
+                    prefix_start_offset = Some(p.offset);
+                }
+
+                match prev {
+                    None => {
+                        prefix_length += 1;
+                        prev = Some(p.offset);
+                    }
+                    Some(offset) => {
+                        if p.offset == offset + 1 {
+                            prefix_length += 1;
+                            prev = Some(p.offset);
+                        } else {
+                            break;
+                        }
+                    }
+                }
+            }
+
+            let prefix_end_offset = prev.expect("known to exist");
+            trace!(
+                "partition {} found a prefix of {:?}",
+                partition_id,
+                prefix_length
+            );
+
+            if prefix_length > flush_min_records {
+                // We have a "large enough" prefix. Lets write it to a file
+                let mut buf = Vec::new();
+                for record in partition.pending.drain(..prefix_length) {
+                    let row = Row::pack(&[
+                        Datum::Int64(record.offset),
+                        Datum::Int64(record.timestamp as i64),
+                        Datum::Bytes(&record.key),
+                        Datum::Bytes(&record.payload),
+                    ]);
+
+                    row.encode(&mut buf);
+                }
+
+                // The offsets we put in this filename are 1-indexed
+                // MzOffsets, so the starting number is off by 1 for something like
+                // Kafka
+                let filename = format!(
+                    "materialize-{}-{}-{}-{}",
+                    self.id,
+                    partition_id,
+                    prefix_start_offset.unwrap(),
+                    prefix_end_offset
+                );
+
+                // We'll write down the data to a file with a `-tmp` prefix to
+                // indicate a write was in progress, and then atomically rename
+                // when we are done to indicate the write is complete.
+                let mut path = self.path.clone();
+                path.push(format!("{}-tmp", filename));
+
+                std::fs::write(&path, buf)?;
+                let final_path = path.with_file_name(filename);
+                std::fs::rename(path, final_path)?;
+                partition.last_persisted_offset = Some(prefix_end_offset);
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub enum PersistenceMetadata {
+    AddSource(GlobalId),
+    DropSource(GlobalId),
+    Shutdown,
+}
+
+pub struct Persister {
+    data_rx: comm::mpsc::Receiver<WorkerPersistenceData>,
+    metadata_rx: std::sync::mpsc::Receiver<PersistenceMetadata>,
+    sources: HashMap<GlobalId, Source>,
```

Comment:
I am pretty convinced that keying sources on GlobalId is wrong in the case where the same source is used multiple times in different dataflows